cli/azd/pkg/tools/azcli/ad.go

```diff
@@ -257,12 +258,26 @@ func (cli *azCli) applyRoleAssignmentWithRetry(
 		}, nil)
 
 		if err != nil {
-			// If the response is a 409 conflict then the role has already been assigned.
 			var responseError *azcore.ResponseError
+			// If the response is a 409 conflict then the role has already been assigned.
 			if errors.As(err, &responseError) && responseError.StatusCode == http.StatusConflict {
 				return nil
 			}
 
+			// If the response is a 403 then the required role is missing.
+			if errors.As(err, &responseError) && responseError.StatusCode == http.StatusForbidden {
+				return retry.RetryableError(
+					fmt.Errorf(
+						"failed assigning role assignment '%s' to service principal '%s' : %w \n%s",
+						*roleDefinition.Name,
+						servicePrincipal.DisplayName,
+						err,
+						output.WithHighLightFormat("Suggested Action: Add `User Access Administrator` or `Owner` azure roles to "+
```

Comment:
I think using `output.WithHighlightFormat` in our errors strings is a bad idea.  It's going to mean that the string returned by this `error` when you call it's `.Error()` method is going to have ANSI escape codes in it.  That's good for logging to the console (as we see in your wonderful screenshot) but would be bad if we write the message into a log file or our telemetry system, because the ANSI escape codes are meaningless there and will just clutter the message.

In addition, from your screenshot, we see that the suggestion ends up getting printed before the TraceID for the error, instead of after, and it feels like the right pattern would be to have the suggestion at the end.

I wonder if we could do something like the following:

Create a type like `ErrorWithSuggestion` that looks something like this:

```golang
type ErrorWithSuggestion struct {
    Suggestion string
    Error error
}

func (es *ErrorWithSuggestion) Error() string {
    return es.Error.Error()
}
```

Then, also implement errors.As and errors.Is for this wrapper type, so it plays well with the rest of our error handling.

Then, in [cobra_builder.go](https://github.com/Azure/azure-dev/blob/4afa01d36c0e9d83db12629de45cd6cd6263cb3e/cli/azd/cmd/cobra_builder.go#L170-L178), add some code that does an `errors.As()` to see if the type can be converted to a `ErrorWithSuggestion`.  If so, print the value of the suggestion string after we write the error message itself (which could include the Trace ID) to the console.  That would allow us to have Suggestion just be an unformatted string, and apply the formatting at the top level, where we do our other error handling.

We could add this ErrorWithSuggestion type to some centralized package and it would give us a way to start easily adding a recommended fix for any error but simply wrapping the existing error in this type with a suggestion and returning it.  If/When that error makes it to the top level of our program, where we do our common error printing, we'll be able to notice there's a recommendation and print it after all the error text.

Interested in @weikanglim's thoughts here, as he has been thinking about our error handling story in general.

Reply:
I'm not certain I understand.

The logic for displaying an error returned by an action already seems centralized to me in `cobra_builder.go` and I think that's the right place for it.

I agree that each action shouldn't need to handle `ErrorWithSuggesion`, but it seems like the structure we have in place today would allow that, simply by adding additional code to `cobra_builder.go` with our existing error handling code.

All things being equal, I have a preference for this more direct style.